base/compiler/ssair/queries.jl

```diff
@@ -20,17 +20,16 @@ function stmt_effect_free(@nospecialize(stmt), src, spvals::SimpleVector)
         ea = e.args
         if head === :call
             f = argextype(ea[1], src, spvals)
-            if isa(f, Const)
-                f = f.val
-            elseif isType(f)
-                f = f.parameters[1]
-            else
-                return false
-            end
+            f = isa(f, Const) ? f.val : isType(f) ? f.parameters[1] : return false
             f === return_type && return true
-            # TODO: This needs significant refinement
-            contains_is(_PURE_BUILTINS, f) || return false
-            return builtin_nothrow(f, Any[argextype(ea[i], src, spvals) for i = 2:length(ea)])
+            if f === _apply
+                f = argextype(ea[2], src, spvals)
+                f = isa(f, Const) ? f.val : isType(f) ? f.parameters[1] : return false
+            end
```

Comment:
The `argextype` of `f` is not the `argextype` from the original expression. But if you're here, doesn't that mean inlining already decided that this `_apply` expression can't be simplified (re-expressed as a call)?

Reply:
Yes, you're probably correct. In particular, if we knew the length of the tuple, inlining would have rewritten it as a call, and if we don't, we also can't really say anything about nothrow.